File: Classes/Communications/StatusSocket.vb
'nexIRC 3.0.31
'05-30-2016 - guideX
Option Explicit On
Option Strict Off
Imports nexIRC.clsIrcNumerics
Imports nexIRC.Modules

Namespace Classes.Communications
    Public Class StatusSocket
        Public Event DataArrival(data As String)
        Private WithEvents socket As AsyncSocket
        Private statusId As Integer
        Private Delegate Sub StringDelegate(ByVal data As String)
        Private Delegate Sub IntegerDelegate(ByVal int As Integer)
        Private Delegate Sub DataArrivalDelegate(ByVal id As Integer, ByVal data As String)
        Private Delegate Sub DisconnectDelegate(ByVal id As Integer, ByVal closeSocket As Boolean)
        Private _invoke As Form
        Private _socketType As SocketType = SocketType.Status

        Public Enum SocketType
            Status = 1
            Ident = 2
        End Enum

        Public WriteOnly Property SetSocketType() As SocketType
            Set(value As SocketType)
                _socketType = value
            End Set
        End Property

        Public WriteOnly Property PassSocket(invokeForm As Form) As AsyncSocket
            Set(value As AsyncSocket)
                _invoke = invokeForm
                socket = value
            End Set
        End Property

        Public Function Connected() As Boolean
                Return socket.Connected
        End Function

        Public Function ReturnLocalIp() As String
                ReturnLocalIp = socket.ReturnLocalIp()
        End Function

        Public Function ReturnLocalPort() As Long
                If (Connected() = True) Then
                    Return socket.ReturnLocalPort
                End If
                Return 0
        End Function

        Public Sub NewSocket(ByVal id As Integer, ByVal form As Form)
                socket = New AsyncSocket
                statusId = id
                _invoke = New Form
                _invoke = form
        End Sub

        Public Sub SendSocket(ByVal data As String)
                If (Connected() = True) Then socket.Send(data & Environment.NewLine)
        End Sub

        Public Sub CloseSocket()
                If Connected() = True Then socket.Close()
        End Sub

        Public Sub ConnectSocket(ByVal ip As String, ByVal port As Long)
                If Connected() = False Then socket.Connect(ip, port)
        End Sub

        Public Sub CouldNotConnect(data As String)
            lStrings.ProcessReplaceString(statusId, eStringTypes.sCOULD_NOT_CONNECT, lStatus.ServerDescription(statusId))
        End Sub

        Private Sub socket_CouldNotConnect(SocketID As String) Handles socket.CouldNotConnect
            Dim couldNotConnectEvent As New StringDelegate(AddressOf CouldNotConnect)
            _invoke.Invoke(couldNotConnectEvent, SocketID)
        End Sub

        Private Sub lSocket_socketConnected(ByVal socketID As String) Handles socket.SocketConnected
            Try
                Dim connectEvent As New IntegerDelegate(AddressOf lStatus.ConnectEvent)
                _invoke.Invoke(connectEvent, statusId)
            Catch ex As Exception
                Throw
            End Try
        End Sub

        Public Sub DataArrivalProc(id As Integer, data As String)
            RaiseEvent DataArrival(data)
        End Sub

        Private Sub lSocket_socketDataArrival(ByVal socketID As String, ByVal socketData As String, ByVal bytes() As Byte, ByVal lByteRead As Integer) Handles socket.SocketDataArrival
            Select Case _socketType
                Case SocketType.Status
                    Dim processDataArrival As New DataArrivalDelegate(AddressOf lProcessNumeric.lIrcNumericHelper.ProcessDataArrival)
                    _invoke.Invoke(processDataArrival, statusId, socketData)
                Case SocketType.Ident
                    Dim processDataArrival As New DataArrivalDelegate(AddressOf DataArrivalProc)
                    _invoke.Invoke(processDataArrival, statusId, socketData)
            End Select
        End Sub

        Private Sub socket_socketDisconnected(ByVal socketId As String) Handles socket.SocketDisconnected
            Try
                Dim disconnectEvent As New DisconnectDelegate(AddressOf lStatus.CloseStatusConnection)
                If Connected() = True Then _invoke.Invoke(disconnectEvent, statusId, False)
            Catch ex As Exception
                'Throw
            End Try
        End Sub
    End Class
End Namespace

File: nexIRC.Business/Sockets/StatusSocket.cs
/*
//nexIRC 3.0.31
//05-30-2016 - guideX
using System;
using System.Windows.Forms;
namespace nexIRC.Business.Sockets {
    public class StatusSocket {
        public event DataArrivalEventHandler DataArrival;
        public delegate void DataArrivalEventHandler(string data);
        private AsyncSocket withEventsField_socket;
        private AsyncSocket socket {
            get { return withEventsField_socket; }
            set {
                if (withEventsField_socket != null) {
                    withEventsField_socket.CouldNotConnect -= socket_CouldNotConnect;
                    withEventsField_socket.SocketConnected -= lSocket_socketConnected;
                    withEventsField_socket.SocketDataArrival -= lSocket_socketDataArrival;
                    withEventsField_socket.SocketDisconnected -= socket_socketDisconnected;
                }
                withEventsField_socket = value;
                if (withEventsField_socket != null) {
                    withEventsField_socket.CouldNotConnect += socket_CouldNotConnect;
                    withEventsField_socket.SocketConnected += lSocket_socketConnected;
                    withEventsField_socket.SocketDataArrival += lSocket_socketDataArrival;
                    withEventsField_socket.SocketDisconnected += socket_socketDisconnected;
                }
            }
        }
        private int statusId;
        private delegate void StringDelegate(string data);
        private delegate void IntegerDelegate(int @int);
        private delegate void DataArrivalDelegate(int id, string data);
        private delegate void DisconnectDelegate(int id, bool closeSocket);
        private Form _invoke;
        private SocketType _socketType = SocketType.Status;
        public enum SocketType {
            Status = 1,
            Ident = 2
        }
        public SocketType SetSocketType {
            set { _socketType = value; }
        }

        public AsyncSocket PassSocket {
            set {
                _invoke = invokeForm;
                socket = value;
            }
        }
        public bool Connected() {
            return socket.Connected;
        }
        public string ReturnLocalIp() {
            return socket.ReturnLocalIp();
        }
        public int ReturnLocalPort() {
            if ((Connected() == true)) {
                return socket.ReturnLocalPort();
            }
            return 0;
        }

        public void NewSocket(int id, Form form) {
            socket = new AsyncSocket();
            statusId = id;
            _invoke = new Form();
            _invoke = form;
        }

        public void SendSocket(string data) {
            if ((Connected() == true))
                socket.Send(data + Environment.NewLine);
        }

        public void CloseSocket() {
            if (Connected() == true)
                socket.Close();
        }

        public void ConnectSocket(string ip, int port) {
            if (Connected() == false)
                socket.Connect(ip, port);
        }

        public void CouldNotConnect(string data) {
            lStrings.ProcessReplaceString(statusId, IrcNumeric.sCOULD_NOT_CONNECT, lStatus.ServerDescription(statusId));
        }

        private void socket_CouldNotConnect(string SocketID) {
            StringDelegate couldNotConnectEvent = new StringDelegate(CouldNotConnect);
            _invoke.Invoke(couldNotConnectEvent, SocketID);
        }

        private void lSocket_socketConnected(string socketID) {
            try {
                IntegerDelegate connectEvent = new IntegerDelegate(lStatus.ConnectEvent);
                _invoke.Invoke(connectEvent, statusId);
            } catch (Exception ex) {
                throw;
            }
        }

        public void DataArrivalProc(int id, string data) {
            if (DataArrival != null) {
                DataArrival(data);
            }
        }

        private void lSocket_socketDataArrival(string socketID, string socketData, byte[] bytes, int lByteRead) {
            switch (_socketType) {
                case SocketType.Status:
                    DataArrivalDelegate processDataArrival = new DataArrivalDelegate(lProcessNumeric.lIrcNumericHelper.ProcessDataArrival);
                    _invoke.Invoke(processDataArrival, statusId, socketData);
                    break;
                case SocketType.Ident:
                    DataArrivalDelegate p = new DataArrivalDelegate(DataArrivalProc);
                    _invoke.Invoke(p, statusId, socketData);
                    break;
            }
        }

        private void socket_socketDisconnected(string socketId) {
            DisconnectDelegate disconnectEvent = new DisconnectDelegate(lStatus.CloseStatusConnection);
            if (Connected() == true)
                _invoke.Invoke(disconnectEvent, statusId, false);
        }
    }
}*/